server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+  - Does not block waiting for verification; the hash computation and verification continue asynchronously.
+- **BlockHashingSessionFactory**
+  - Creates new BlockHashingSession instances, provides them with a ExecutorService.
+- **BlockHashingSession**
+  - Holds all necessary block data.
+  - Accepts block items incrementally. (continues to compute them asynchronously)
+  - Once the block_proof is provided, finalizes the hash computation asynchronously.
+  - After computing the final hash, calls SignatureVerifier for verification.
+- **SignatureVerifier**
+  - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).
+  - Calls BlockStatusManager with the verification result.
+- **BlockStatusManager**
+  - Receives verification results from SignatureVerifier.
+  - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.
+
+## Design
+
+1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.
+2. When the block_header is detected, the `VerificationHandler` creates a `BlockHashingSession` using the `BlockHashingSessionFactory`.
+3. The `BlockHashingSession` accepts subsequent block items incrementally.
+4. Once the block_proof is received, the `BlockHashingSession` calls `completeHashing()` to finalize the hash computation.
+5. Upon completion of computing the final block hash, the `BlockHashingSession` calls the `SignatureVerifier` to verify the signature.
+6. The `SignatureVerifier` compares the computed hash to the hash implied by the signature using the public key.
+7. If the verification fails, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as SIGNATURE_INVALID.
+8. If the verification succeeds, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as VERIFIED.
+9. The `BlockStatusManager` triggers any necessary recovery or follow-up processes depending on the verification result.
+
+Sequence Diagram:
+```mermaid
+sequenceDiagram
+    participant U as UnverifiedRingBuffer
+    participant V as VerificationHandler
+    participant F as BlockHashingSessionFactory
+    participant S as BlockHashingSession
+    participant SV as SignaturesequenceDiagram
+    participant U as UnverifiedRingBuffer
+    participant V as VerificationHandler
+    participant F as BlockHashingSessionFactory
+    participant S as BlockHashingSession
+    participant SV as SignatureVerifier
+    participant BSM as BlockStatusManager
+
+        
+    U->>V: (1) onBlockItemsReceived(List<BlockItem>)        
+    
+    alt (2) Detects block_header
+    
+    V->>F: createSession(initialBlockItems, executorService, signatureVerifier)
+    Note over S: New instance of BlockHashingSession created
+    F-->>V: returns BlockHashingSession (S)
+    V-->>U: Returns without blocking
```

Comment:
Note, minimizing time on ring buffer threads is one part of making this performance equation work.
The other is possibly changing clients to use a blocking poll rather than a callback/reactive approach.  Certainly not what we need/want to do at the moment, but it may well be worth thinking about how that would work in a future update.

Reply:
This matches how I understood it, and our slack discussion was helpful.

However, I think we should at least consider Jasper's suggestion that we try polling and/or blocking polling for at least some of the Ring Buffer clients.  We may be able to further reduce thread transitions and avoid some of the thread-lock possibilities inherent in the "push" option.